pkg/eestream/encode.go

```diff
@@ -38,123 +40,309 @@ type ErasureScheme interface {
 }
 
 type encodedReader struct {
-	r               io.Reader
-	es              ErasureScheme
-	cv              *sync.Cond
-	inbuf           []byte
-	outbufs         [][]byte
-	piecesRemaining int
-	err             error
+	ctx    context.Context
+	cancel context.CancelFunc
+	r      io.Reader
+	es     ErasureScheme
+	min    int // minimum threshold
+	opt    int // optimum threshold
+	inbuf  []byte
+	eps    map[int](*encodedPiece)
+	mux    sync.Mutex
+	start  time.Time
+	done   int // number of readers done
 }
 
-// EncodeReader will take a Reader and an ErasureScheme and return a slice of
-// Readers
-func EncodeReader(r io.Reader, es ErasureScheme) []io.Reader {
+// EncodeReader takes a Reader and an ErasureScheme and return a slice of
+// Readers.
+//
+// min is the minimum threshold - the minimum number of erasure pieces that
+// are completely stored. If set to 0, it will be reset to the TotalCount
+// of the ErasureScheme.
+// opt is the optimum threshold - the optimum number of erasure pieces that
+// are completely stored. If set to 0, it will be reset to the TotalCount
+// of the ErasureScheme.
+// mbm is the maximum memory (in bytes) to be allocated for read buffers. If
+// set to 0, the minimum possible memory will be used.
+//
+// When the minimum threshold is reached a timer will be started with another
+// 1.5x the amount of time that took so far. The Readers will be aborted as
+// soon as the timer expires or the optimum threshold is reached.
+func EncodeReader(ctx context.Context, r io.Reader, es ErasureScheme,
+	min, opt, mbm int) []io.Reader {
+	if min == 0 {
+		min = es.TotalCount()
+	}
+	if opt == 0 {
+		opt = es.TotalCount()
+	}
+	err := checkParams(es, min, opt, mbm)
 	er := &encodedReader{
-		r:       r,
-		es:      es,
-		cv:      sync.NewCond(&sync.Mutex{}),
-		inbuf:   make([]byte, es.DecodedBlockSize()),
-		outbufs: make([][]byte, es.TotalCount()),
+		r:     r,
+		es:    es,
+		min:   min,
+		opt:   opt,
+		inbuf: make([]byte, es.DecodedBlockSize()),
+		eps:   make(map[int](*encodedPiece), es.TotalCount()),
+		start: time.Now(),
 	}
+	er.ctx, er.cancel = context.WithCancel(ctx)
 	readers := make([]io.Reader, 0, es.TotalCount())
 	for i := 0; i < es.TotalCount(); i++ {
-		er.outbufs[i] = make([]byte, 0, es.EncodedBlockSize())
-		readers = append(readers, &encodedPiece{
-			er: er,
-			i:  i,
-		})
+		er.eps[i] = &encodedPiece{
+			er:  er,
+			err: err,
+		}
+		er.eps[i].ctx, er.eps[i].cancel = context.WithCancel(er.ctx)
+		readers = append(readers, er.eps[i])
+	}
+	if err != nil {
+		return readers
```

Comment:
the thing with the `err` seems really clever, but it's kind of surprising and hard to read, and i'm surprised we don't just change EncodeReader to return `([]io.Reader, error)`